### PR TITLE
FIR: report VARARG_OUTSIDE_PARENTHESES

### PR DESCRIPTION
**Pull request by tgeng**

### CODE REVIEW

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnostics.kt [hunk 1]
```diff
@@ -534,7 +539,7 @@ sealed class KtFirDiagnostic<PSI: PsiElement> : KtDiagnosticWithPsi<PSI> {
         override val diagnosticClass get() = NonFinalMemberInObject::class
     }
 
-    abstract class ManyCompanionObjects : KtFirDiagnostic<PsiElement>() {
+    abstract class ManyCompanionObjects : KtFirDiagnostic<KtObjectDeclaration>() {
```
Comment (punzki) [inline idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnostics.kt:542]: Why is this change necessary?

Comment (Ebiabi) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnostics.kt:542]: What is wrong bro?

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt [hunk 2]
```diff
@@ -166,6 +166,8 @@ object DIAGNOSTICS_LIST : DiagnosticList() {
         val INAPPLICABLE_LATEINIT_MODIFIER by error<FirSourceElement, KtModifierListOwner>(PositioningStrategy.LATEINIT_MODIFIER) {
             parameter<String>("reason")
         }
+
+        val VARARG_OUTSIDE_PARENTHESES by error<FirSourceElement, PsiElement>()
```
Comment (jsjeon) [inline compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:170]: s/PsiElement/KtExpression/g ?

Comment (jsjeon) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:170]: I forgot to answer this: https://github.com/JetBrains/kotlin/pull/4143#discussion_r585285315 So, please double-check on what type the counterpart of old FE's diagnostic is reported.

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnostics.kt (hunk 1 above)
Comment (tgeng) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnostics.kt:542]: Hmm. Weird. I actually didn't change this and this is changed in https://github.com/JetBrains/kotlin/commit/6fee46316eb9836a13a8367c001d9122fff5974f. Let me sync master again and re-push.

Comment (punzki) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnostics.kt:542]: Huh, ok I guess the generator was not run then. 🤷

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnosticsImpl.kt [hunk 3]
```diff
@@ -617,6 +618,13 @@ internal class InapplicableLateinitModifierImpl(
     override val firDiagnostic: FirPsiDiagnostic<*> by weakRef(firDiagnostic)
 }
 
+internal class VarargOutsideParenthesesImpl(
+    firDiagnostic: FirPsiDiagnostic<*>,
+    override val token: ValidityToken,
+) : KtFirDiagnostic.VarargOutsideParentheses(), KtAbstractFirDiagnostic<PsiElement> {
```
Comment (jsjeon) [inline idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDiagnosticsImpl.kt:624]: I wonder this could be more specific, i.e., `KtExpression` too.

File: compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt (hunk 2 above)
Comment (tgeng) [reply compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:170]: Thanks for the explanation!

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually 👍
